roadmap/implementers-guide/src/runtime/disputes.md

```diff
@@ -17,90 +19,350 @@ We account for these requirements by having the validity module handle two kinds
+      1. slash for double vote
+    1. count vote as part of stored `Votes`
+    1. check for super majority
+      1. transplant resolution and slashes
+
+
+* `on_unavailability_timeout(hash: Hash)`
+  1. no super majority was reached
+    1. iff both parties got > 1/3 of the votes, BUG
+    1. else, slash all voting validators
+  1. ...
 
-If the candidate was not made available on another fork of the relay chain, the availability process will time out and the disputing validator will be slashed on this fork. The escalation used by the validator(s) can be replayed onto other forks to lead the wrongly-escalating validator(s) to be slashed on all other forks as well. We assume that the adversary cannot censor validators from seeing any particular forks indefinitely
 
-> TODO: set the availability timeout for this accordingly - unlike in the inclusion pipeline we are slashing for unavailability here!
+* `shift_session(current: SessionIndex, previous: SessionIndex)`
+  1. obtain the current sessions validator set
+  2. copy that data back to the previsou session
+  3. cancel all ongoing disputes of the now obsolete session
 
-If the availability process passes, the remote dispute is ready to be included on this chain. As with the local dispute, validators self-select based on a VRF. Given that a remote dispute is likely to be replayed across multiple forks, it is important to choose a VRF in a way that all forks processing the remote dispute will have the same one. Choosing the VRF is important as it should not allow an adversary to have control over who will be selected as a secondary approval checker.
+* `on_session_change()`
+  1. obtain current session index
+  1. obtain previous session index
+  1. call `shift_session`
+
+## Off-chain
 
-After enough validator self-select, under the same escalation rules as for local disputes, the Remote dispute will conclude, slashing all those on the wrong side of the dispute. After concluding, the remote dispute remains open for a set amount of blocks to accept any further proof of additional validators being on the wrong side.
+* `craft_slashing_inherent(to_slash: &[(SessionIndex, ValidatorIndex)]) -> Inherent`
+  > TODO: ...
+
+* `transplant(slashing_inherent : Inherent)`
+
+  > TODO: pending `ActiveLeaves` tracking in `ChainAPI` subsystem
+
+  1. query `active_heads` set
+  1. for each `active_head` in `active_heads`
+    1. queue a transaction with the `slashing_inherent`
+
+
+# Open Questions
+
+  - what happens with open disputes on session / era change?
+  - > TODO: validator-dispute could be instead replaced by a fisherman w/ bond
+  - > TODO: Given that a remote dispute is likely to be replayed across multiple forks, it is important to choose a VRF in a way that all forks processing the remote dispute will have the same one. Choosing the VRF is important as it should not allow an adversary to have control over who will be selected as a secondary approval checker.
```

Comment:
@rphmeier I would be in favour of dropping the VRF references here, it does not really belong here and is a bit distracting.

Reply:
Agreed, approval assignment VRFs play no role in disputes.  It's possible this comment should be moved somewhere under approvals, not sure. 

As for session/era changes, we want disputes to "always" run to conclusion, and we slash anyone voting the wrong way, even if not in the current validator set and thus their vote does not count, but we still must select exactly which validator set gets used when.  We'd ideas when we last talked, but not perfectly well backed in my mind. 

I think the choices go:  We keep the set from whenever the dispute gets posted on-chain. Vs. We repost on-chain any dispute that runs out the clock, so the new validator set takes over.  We thought Rob's intuition sounded useful here because grandpa must run into similar issues.

Reply:
> Agreed, approval assignment VRFs play no role in disputes. It's possible this comment should be moved somewhere under approvals, not sure.

I am going to drop it.

> As for session/era changes, we want disputes to "always" run to conclusion, and we slash anyone voting the wrong way, even if not in the current validator set and thus their vote does not count, but we still must select exactly which validator set gets used when. We'd ideas when we last talked, but not perfectly well backed in my mind.

Is `current` here would imply the session that was present at the time of dispute (which is part of the local dispute gossip message) plus the following session in case of a session change?

> I think the choices go: We keep the set from whenever the dispute gets posted on-chain. Vs. We repost on-chain any dispute that runs out the clock, so the new validator set takes over. We thought Rob's intuition sounded useful here because grandpa must run into similar issues.

The thinking here was to use the historical slashing pallet, which stores the last k validator sets to the off-chain.

Reply:
Sounds good.  I meant which ever one we pick to settle the dispute, so current might not be the best word.

Reply:
We shouldn't need re-posting as long as we expect that validators will remain active for a few sessions afterwards.